### DISCUSSION TITLE
Cannot destructure property 'stanza' of 'ctx' as it is undefined

### DISCUSSION THREAD

**Discussion by Domingez43** [xmppjs/xmpp.js#982] [Q&A]
Hello, im trying to connect to the FCM endpoint and i still get this error:

`node_modules\@xmpp\stream-features\index.js:17
      const { stanza } = ctx;
              ^
TypeError: Cannot destructure property 'stanza' of 'ctx' as it is undefined.
    at Client.starttls ←[90m(C:\Users\ASUS\OneDrive\Počítač\BACKENDIK\←[39mnode_modules\←[4m@xmpp←[24m\stream-features\index.js:17:15←[90m)←[39m
    at Object.<anonymous> ←[90m(C:\Users\ASUS\OneDrive\Počítač\BACKENDIK\←[39mmessage.js:36:6←[90m)←[39m
←[90m    at Module._compile (node:internal/modules/cjs/loader:1218:14)←[39m
←[90m    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)←[39m
←[90m    at Module.load (node:internal/modules/cjs/loader:1081:32)←[39m
←[90m    at Module._load (node:internal/modules/cjs/loader:922:12)←[39m
←[90m    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)←[39m
←[90m    at node:internal/main/run_main_module:23:47←[39m

Node.js v18.13.0`

Here is my code, its basically edited code from readme. I tried reinstalling packages, but nothing work for me 

```const admin = require("firebase-admin");
const { client, xml } = require("@xmpp/client");
const serviceAccount = require("./bachelor-thesis-firebase-adminsdk.json");
const debug = require("@xmpp/debug");


const projectId = 'bachelor-thesis-f9301';
const serverKey = 'XXX';
const senderId = 184051903478;

const fcmServer = `fcm-xmpp.googleapis.com:5236`;

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const xmpp = client({
    service: 'xmpps://fcm-xmpp.googleapis.com:5236',
    username: `${senderId}@${fcmServer}`,
    password: 'XXXX',
});

debug(xmpp, true);

xmpp.on("error", (err) => {
  console.error(err);
});

xmpp.on("offline", () => {
  console.log("offline");
});

xmpp.on("stanza", async (stanza) => {
  if (stanza.is("message")) {
    await xmpp.send(xml("presence", { type: "unavailable" }));
  }
});

xmpp.starttls().catch(console.error);```

**Comment by sonnyp** [comment]
Never used/tested xmpp.js on windows.

Can you try in WSL? 

I suggest looking at xmpp.js code where it is throwing and trying to found out what's going on.

**Reply by Domingez43** [reply]
I tried it in WSL and also on MacOS same error when trying to use tls connection. When Im not using TLS it connects, but i get error: "self-signed certificate". So I guess Firebase require TLS to connect to its endpoint.

It seems like middleware here in xmpp/stream-features returns nothing. 

```const route = require("./route");

module.exports = function streamFeatures({ middleware }) {
  middleware.use(route());

  function use(name, xmlns, handler) {
    return middleware.use((ctx, next) => {
      const { stanza } = ctx;
      if (!stanza.is("features", "http://etherx.jabber.org/streams"))
        return next();
      const feature = stanza.getChild(name, xmlns);
      if (!feature) return next();
      return handler(ctx, next, feature);
    });
  }
  return {
    use,
  };
};
